### PR TITLE
deps: use Python 3.8

### PR DESCRIPTION
**Pull request by epuertat**
Reasons:
* Making dependencies on Python minor version numbers explicit ([PEP-20](https://www.python.org/dev/peps/pep-0020/): explicit is better than implicit): Ceph packages don't explicitly require a minor Python version, but they actually rely on 3.6 features (f-strings just to name one).
* Python 3.6 reaches End of Life this Dec '21 (3.8 EoL is [Oct '24](https://www.python.org/dev/peps/pep-0569/)).
* Python 3.8 provides new features that might help write more efficient and solid code:
  * Improved `typing` support
  * `dataclasses`
  * `async`
  * An `statistics` package
  * And other features [[1]](https://docs.python.org/3/whatsnew/3.8.html) [[2]](https://docs.python.org/3/whatsnew/3.7.html)
* Performance-wise: Python 3.7/3.8 brought [dramatic performance improvements](https://speed.python.org/comparison/?exe=12%2BL%2B3.6%2C12%2BL%2B3.8%2C12%2BL%2B3.9%2C12%2BL%2Bmaster&ben=616%2C617%2C618%2C619%2C620%2C621%2C622%2C623%2C624%2C625%2C626%2C627%2C628%2C629%2C630%2C631%2C632%2C680%2C633%2C634%2C635%2C636%2C637%2C638%2C639%2C640%2C641%2C642%2C643%2C644%2C645%2C646%2C647%2C648%2C681%2C649%2C650%2C651%2C652%2C653%2C654%2C655%2C656%2C657%2C658%2C659%2C660%2C661%2C682%2C662%2C663%2C664%2C665%2C666%2C667%2C669%2C668%2C670%2C671%2C672%2C673%2C674%2C675%2C678%2C677%2C676%2C679&env=1%2C2&hor=true&bas=none&chart=stacked+bars):
image

Other changes:
* ceph.spec.in: use `python3_version` and `python3_version_nodots` [Fedora/EPEL RPM macros](https://docs.fedoraproject.org/en-US/packaging-guidelines/Python/#_macros). Added custom `python3_version_major`.

Check distro support:
- [x] CentOS 8
- [x] Ubuntu >20.04
- [x] OpenSUSE Leap 15.3

Fixes: https://tracker.ceph.com/issues/51319
Fixes: https://tracker.ceph.com/issues/56945
Signed-off-by: Ernesto Puerta <epuertat@redhat.com>

## Checklist
- [x] References tracker ticket
- [x] Updates documentation if necessary
- [x] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: CMakeLists.txt
Comment (sebastian-philipp) [inline CMakeLists.txt:471]: I don't have 3.8 installed on my system. Is there a way to make exclude 3.5 without mandating versions that are not installed?

Comment (tchaikov) [reply CMakeLists.txt:471]: i don't think wiring the default value to 3.8 helps unless we depends on 3.8 and up.

File: ceph.spec.in
Comment (tchaikov) [inline ceph.spec.in:118]: does RHEL8 and CentOS8 offer python3.8 ? IIUC, they only offer python 3.6.8. see https://centos.pkgs.org/8/centos-baseos-x86_64/python3-libs-3.6.8-37.el8.i686.rpm.html

File: ceph.spec.in
Comment (tchaikov) [inline ceph.spec.in:229]: i don't understand this change. and i cannot find the explanation in the commit message or in tracker ticket.

File: CMakeLists.txt
Comment (epuertat) [reply CMakeLists.txt:471]: > I don't have 3.8 installed on my system. Is there a way to make exclude 3.5 without mandating versions that are not installed?

@sebastian-philipp but if we declare the 3.8 dep it'll become installed. Is there any specific reason not to want 3.8 installed?

Comment (epuertat) [reply CMakeLists.txt:471]: > i don't think wiring the default value to 3.8 helps unless we depends on 3.8 and up.

Just to mention a single change: f-strings are not supported in 3.5, so we should at least indicate that we require 3.6 or above (which, on the other hand, is nothing really adventurous given that 3.5 end of life finished almost a year ago and 3.5 EOL is Dec this year).

The reason to go with 3.8 is using improved typing constructs and other basic helpers as `dataclasses`.

File: ceph.spec.in
Comment (epuertat) [reply ceph.spec.in:118]: I'm not sure about RHEL, but [CentOS 8 has support via AppStream repo](https://pkgs.org/search/?q=python3.8).

File: ceph.spec.in
Comment (epuertat) [reply ceph.spec.in:229]: @tchaikov sure, I'll update the PR to comment my changes. In this very case, the `python3_pkgversion` macro is not used in the RPM ecosystem, while the [`python3_version_nodots` is a very well established macro](https://docs.fedoraproject.org/en-US/packaging-guidelines/Python/).

File: ceph.spec.in
Comment (tchaikov) [reply ceph.spec.in:118]: i assume that we need to enable the AppStream repo for using it? is all the dependencies available in app stream repo?

File: CMakeLists.txt
Comment (tchaikov) [reply CMakeLists.txt:471]: > > i don't think wiring the default value to 3.8 helps unless we depends on 3.8 and up.
> 
> @tchaikov just to mention a single change: f-strings are not supported in 3.5, so we should at least indicate that we require 3.6 or above (which, on the other hand, is nothing really adventurous given that 3.5 end of life finished almost a year ago and 3.6 EOL is Dec this year).

i agreed that depending on 3.6 is the way to go. that's actually one of the reasons why i pushed the xenial to bionic change.

please note, my concern is not why 3.6 but why 3.8.

File: ceph.spec.in
Comment (tchaikov) [reply ceph.spec.in:229]: i was looking at https://fedoraproject.org/wiki/PackagingDrafts:Python3EPEL and was referencing lots of actively maintained python packages, for instance

- https://src.fedoraproject.org/rpms/python-requests/blob/rawhide/f/python-requests.spec
- https://src.fedoraproject.org/rpms/python-setuptools/blob/rawhide/f/python-setuptools.spec

so i am not convinced that `python3_pkgversion` is not used n the RPM ecosystem.

**Review by tchaikov — CHANGES_REQUESTED** [review]

File: ceph.spec.in
Comment (tchaikov) [reply ceph.spec.in:229]: and this change should be put into a separate commit.

File: ceph.spec.in
Comment (epuertat) [reply ceph.spec.in:118]: @tchaikov I just tried this with the [oficial CentOS8 container image](https://hub.docker.com/_/centos):
```sh
> docker run --rm -ti centos:8 bash
# yum install -y python3.8
...
Installed:
  python38-3.8.6-3.module_el8.4.0+665+abc3a503.x86_64                                                         
  python38-libs-3.8.6-3.module_el8.4.0+665+abc3a503.x86_64                                                    
  python38-pip-19.3.1-1.module_el8.4.0+647+0ba99ce8.noarch                                                    
  python38-pip-wheel-19.3.1-1.module_el8.4.0+647+0ba99ce8.noarch                                              
  python38-setuptools-41.6.0-4.module_el8.4.0+647+0ba99ce8.noarch                                             
  python38-setuptools-wheel-41.6.0-4.module_el8.4.0+647+0ba99ce8.noarch                                       

Complete!
```
So it seems that the AppStream repo is enabled by default.

File: CMakeLists.txt
Comment (jecluis) [reply CMakeLists.txt:471]: Personally I've found a few things that are of interest in python 3.8, but they mainly revolve around `asyncio` improvements. On top of that, I find having type annotation forward references is incredibly useful. And, as @epuertat mentioned, 3.6 is EOL at end of this year, and moving on to python 3.8 makes sense just because of that, should there be no particular constraint.

File: ceph.spec.in
Comment (tchaikov) [reply ceph.spec.in:118]: i see. just read more about AppStream repo. yeah, it seems that it is enabled by default. but i am not sure what the "quality of service" we can expect from it.

1. are all python packages available we need in BaseOS also available in AppStream?
2. what about their lifecycles?
3. do they also get the backport and fixes?

File: ceph.spec.in
Comment (epuertat) [inline ceph.spec.in:453]: @tchaikov what do you think about this one?

I wasn't sure when I added it: it affects the `base` package. On one hand I didn't want to impose a Ceph-wide dependency on Python 3.8, but on the other hand, it also relies on `python3-setuptools` which will in turn depend on `python3x`, so this makes the dependency more explicit. I personally feel that it's more dangerous to allow coexisting versions of Python on a mono-repo (and assume part of our code base is still 3.6 compliant, without making it clear which ones).

File: ceph.spec.in
Comment (epuertat) [reply ceph.spec.in:229]: Thanks for the reference @tchaikov . My main concern with the `python3_pkgversion` is that the name is not very self-describing, ulinke `python3_version`, `python3_version_nodots` or `python3_version_major` (which follows the [CMake convention](https://cmake.org/cmake/help/v3.14/module/FindPython3.html)).

Why should this be in a separate commit?

Comment (tchaikov) [reply ceph.spec.in:229]: > Thanks for the reference @tchaikov . My main concern with the `python3_pkgversion` is that the name is not very self-describing, ulinke `python3_version`, `python3_version_nodots` or `python3_version_major` (which follows the [CMake convention](https://cmake.org/cmake/help/v3.14/module/FindPython3.html)).
> 

IMHO, pkgversion is a good name in this context, in which it specifies the python version of a python package.

i don't understand why we need to follow the CMake convention here.  FindPython3 module follows the convention of [find_package()](https://cmake.org/cmake/help/latest/command/find_package.html#version-selection), so `find_package(Python3 ...)` can work as expected.

> Why should this be in a separate commit?

because i think it is orthogonal to "use Python 3.8".

File: ceph.spec.in
Comment (tchaikov) [reply ceph.spec.in:453]: yeah, the only python tool packaged by `ceph-base` is the `ceph` cli tool once #41999 gets merged. but `ceph-base` in turn depends on `ceph-common` which packages a bunch of ceph python modules, like

- ceph-argparse
- ceph-common

and python bindings for librados, libcephfs and librbd. all of them are build and installed for a certain python version. so yes, it should depend on a certain python runtime versioned with minor version number.

when it comes to `python3-setuptool`, i moved it to ceph-osd in #41999 as a cleanup. not sure what maintainers of ceph-volume think about the python3.8 dependencies. if we can go further by extracting ceph-volume into its own package, we don't need to worry about this problem anymore. but since `ceph-osd` depends on `ceph-base`, it's non-trivial to make the dependency from ceph-volume to "python3" but not "python3.8" explicit. and as you put, it might not be a best practice to *use* multiple python runtime in a deployment. i concur with you on this, as it'd be a pain for testability and maintainability, even it's completely technically viable and supported by python.

in short, 

- i'd avoid having both `python3-setuptools` and `python38` as runtime dependencies of a single package. 
- if ceph-volume is extracted into its own package, and it does not depend on other ceph python modules. it can depend on `python3-setuptools` or `python38-setuptools`. but if we want to use `python3-setuptools`, we have to test both `python36-setuptools` and `python38-setuptools`.

File: ceph.spec.in
Comment (epuertat) [reply ceph.spec.in:229]: > IMHO, pkgversion is a good name in this context, in which it specifies the python version of a python package.

@tchaikov my main concern with that name is that it doesn't specify whether that's the major number only (as it was so far), the complete version with dots or without (as it should be, at least for CentOS and OpenSUSE `python38` and other ABI-bound packages).

> because i think it is orthogonal to "use Python 3.8".
Do you mean replacing the `python3_pkgversion` with `python3_version_nodots`?

I don't think so: maybe it's not compulsory, but you still need 2 macros to cope with packages bound to the 3.8 ABI (e.g.: `python38-Cython`) and pure source code packages (e.g.: `python3-cherrypy`). If you prefer to keep `python3_pkgversion`, which one should be it? The major-version one or the full nodots? That's why I preferred having a self-descriptive macro. What's your preference?

File: ceph.spec.in
Comment (epuertat) [reply ceph.spec.in]: > i see. just read more about AppStream repo. yeah, it seems that it is enabled by default. but i am not sure what the "quality of service" we can expect from it.

`RHEL 8 content is distributed through the two main repositories: BaseOS and AppStream.` [[ref]](https://docs.centos.org/en-US/8-docs/managing-userspace-components/assembly_using-appstream/)

> 1. are all python packages available we need in BaseOS also available in AppStream?

It has even more: BaseOS has 87 python3-* packages, while AppStream has 267.

> 2. what about their lifecycles?
> 3. do they also get the backport and fixes?

> "New versions of most application streams will become available as needed during the full 10 years of the Red Hat Enterprise Linux 8 life. Lifecycles will align to the upstream support life or to product needs as required. Multiple versions of the same application stream may be available at the same time with overlapping lives. Rolling streams for fast-evolving products have only a single version, the newest, available. If a non-security related bug is discovered in a rolling stream, we will work to address it in the next scheduled release. The release cadence will be selected based on what makes sense for the product. Rolling streams will be available and supported for the full life of Red Hat Enterprise Linux major releases."

 [[ref]](https://access.redhat.com/support/policy/updates/rhel8-app-streams-life-cycle)

File: ceph.spec.in
Comment (epuertat) [reply ceph.spec.in]: Fixed.

File: ceph.spec.in
Comment (tchaikov) [reply ceph.spec.in:230]: i don't think we should not differentiate the python packages which contains binaries from the ones which are implemtented in "pure python". as  both of them are packaged and installed for a certain python x.y version, instead of a certain major version. take `python3-cherrypy` as an example
```console
$ rpm -q --provides python3-cherrypy
python3-cherrypy = 18.4.0-1.el8
python3.6dist(cherrypy) = 18.4.0
python3dist(cherrypy) = 18.4.0
```

the reasons why i think `python3_pkgversion` is the right choice is not completely based on by my personal preferences or taste:

1. to follow related package guide line: https://fedoraproject.org/wiki/PackagingDrafts:Python3EPEL
  
   > All references to python3 (except in macro names) need to be replaced with python%{python3_pkgversion}.

2. to follow examples. as it is used by lots of actively maintained python packages. see 
   1. https://src.fedoraproject.org/rpms/python-setuptools/blob/rawhide/f/python-setuptools.spec
   2. https://src.fedoraproject.org/rpms/python-requests/blob/rawhide/f/python-requests.spec

in general, python package maintainers should not hardwire a python package to a certain python3 minor version. python3 means **the** python3.6 in EL8, and it is python3.9 in fedora33. `python3_pkgversion` is used to encapsulate this information, and i **think** it serves this purpose well. i am not an expert on the rpm packaging, and i don't feel comfortable commenting on if other macros are superior or better than `python3_pkgversion` . it is but my 2 cents.

> Do you mean replacing the `python3_pkgversion` with `python3_version_nodots`?

yes. and i don't recommend `python3_version_nodots`. but if you insist, i'd suggest do this in two steps:

1. to replace the `python3_pkgversion` with `python3_version_nodots`. and accompany it with explanations. so far the commit message still does not explain why we need this change. and your explanations in the comments have been changing.
2. to set the default values of the rpm macros from "3" to "3.8". please note, this does not change the dependencies from 3 to 3.8, as the behavior depends on the existence of `python-srpm-macros`, `python-rpm-macros`, `python36-rpm-macros` and `python38-rpm-macros` .

instead i'd suggest:

1. install `python38-rpm-macros` in `install-deps.sh`. with which,  we have:

  ```console
  $ rpm --eval '%{python3_pkgversion}'
  38
  ```
2. replace `python3-devel` with `python38-devel` in `install-deps.sh`
3. update https://github.com/ceph/ceph-build/blob/master/scripts/build_utils.sh#L1396
4. make sure readthedoc build does not break. see https://readthedocs.org/projects/ceph/builds/14110525/ for an example, where python3.7 instead of python3.8 is used, but `typing.Literal` was not available in python3.7.

Comment (tchaikov) [reply ceph.spec.in:230]: RtD change is created at #42577

File: ceph.spec.in
Comment (tchaikov) [reply ceph.spec.in]: split ceph-volume pending on review at #42465

**Comment by tchaikov** [comment]
@epuertat is this change still relevant ? i will help migrating to python 3.8 if you are fine with this.

File: ceph.spec.in
Comment (tchaikov) [reply ceph.spec.in]: BTW, we will have to build python bindings for both python3.8 and python3.6, so we don't break the existing users of these python bindings.

**Comment by epuertat** [comment]
> @epuertat is this change still relevant ? i will help migrating to python 3.8 if you are fine with this.

@tchaikov I'm more than fine with any help with this! This will become more and more relevant as long as we approach 3.8 EoL by end of this year.

**Comment by jecluis** [comment]
@epuertat any idea what's needed to move this along?

**Comment by epuertat** [comment]
> @epuertat any idea what's needed to move this along?

@jecluis apart from rebasing... Based on [this comment](https://github.com/ceph/ceph/pull/41979#discussion_r680537806) from @tchaikov, we'd need to modify our build pipeline to create 2 different set of packages for 3.8 and 3.6....

But to make things even funnier, I just tested with our 4 major target distros, and found that:
* Both OpenSUSE Leap 15.3 and Debian 11 don't support 3.8, but just 3.9.
* Both CentOS and Ubuntu provide support for both 3.8 and 3.9.

This is the detail on the testing with the following with docker images (`centos:8`, `ubuntu:20.04`, `opensuse/leap:15`, `debian:11`):

Distro   | Release | Default Python 3 | 3.8 available
-------- | ------- | ---------------- | -------------
CentOS   | 8       | 3.6.8            | 3.8.6</br>`dnf module -y install python38` ([AppStream](https://docs.centos.org/en-US/8-docs/managing-userspace-components/assembly_using-appstream/))</br>`python3 --version`: Python 3.8.6
Ubuntu   | 20.04   | **3.8.10**       | By default</br>`apt install python3`
OpenSUSE | 15.3    | 3.6.13           | No 3.8, 15.3 will directly support [3.9](https://software.opensuse.org/package/python38).
Debian   | 11      | **3.9.2**        | No 3.8 support, only 3.9.

Summarizing, if we want to support the above distro matrix, we should better go for 3.9 :/ Good thing is that the 3.9 will give us an extra year (EoL by 2025) and adds [some features](https://docs.python.org/3/whatsnew/3.9.html), like `Annotations`.

**Comment by epuertat** [comment]
Harsher than initially expected. I barely started with CentOS and the main issue is that Python packages are installed in a specific `/usr/lib/python3.6/site-packages/...` location, even if they're cross-version compatible (same as with OpenSUSE, and just the opposite as Debian/Ubuntu distros where things go in `/usr/lib/python3/dist-packages`).

It can be easily hacked with a `PYTHONPATH=/usr/lib/python3.6/site-packages/ python3` though :(

Tempted of doing [this](https://xkcd.com/1654/).

**Comment by epuertat** [comment]
Package | CentOS 8 | OpenSUSE 15.3 | Ubuntu 20.04 | Debian 11 | FreeBSD 13
------- | -------- | ------------- | ------------ | --------- | ----------
Python 3 | `python39` | `python39` | `python3.8` :x:</br>(No Cython for 3.9) | `python3` (3.9) | `python39`
Python-only packages</br>(e.g.: Prettytable) | `python39-prettytable` :x:</br>(in Fedora 33/34 :heavy_check_mark:) | `python39-PrettyTable` :x:</br>(in Tumbleweed :heavy_check_mark:) | `python3-prettytable`</br>(3.x cross-compatible) | `python3-prettytable`</br>(3.x cross-compatible) | `py37-prettytable` :x:

**Comment by ktdreyer** [comment]
Red Hat's Python team will continue to maintain Python 3.6 in RHEL 8 (main EOL is May 31, 2029) and CentOS Stream 8 (EOL is May 31 2024).

We should consider dropping CentOS 8 support for Ceph in `main`. CentOS 9 ships Python 3.9 as the default Python version.

**Comment by cdeshmukh** [comment]
@epuertat we were also looking for python-3.8(.13 as latest) upgrade, for reasons you had listed in PR description, plus in [#56945 issue](https://tracker.ceph.com/issues/56945). Additionally, python-3.6 also has some [security (CVEs) issues](https://www.cvedetails.com/vulnerability-list.php?vendor_id=10210&product_id=18230&version_id=&page=1&hasexp=0&opdos=0&opec=0&opov=0&opcsrf=0&opgpriv=0&opsqli=0&opxss=0&opdirt=0&opmemc=0&ophttprs=0&opbyp=0&opfileinc=0&opginf=0&cvssscoremin=0&cvssscoremax=0&year=0&month=0&cweid=0&order=1&trc=86&sha=d6f3b397c4f4a621c374ac6b27a7ce2b00fb2599) which asks for this upgrade strongly. 


> @epuertat is this change still relevant ? i will help migrating to python 3.8 if you are fine with this.
@tchaikov I'm more than fine with any help with this! This will become more and more relevant as long as we approach 3.8 EoL by end of this year.

After looking at this thread, saw much of the work and discussion has had happened already and was hopeful. Though after reading rest of the comments (below), seems additional packaging challenges would be added, which will be worth too i think, though would be more bulky, right?
https://github.com/ceph/ceph/pull/41979#issuecomment-919397849
https://github.com/ceph/ceph/pull/41979#issuecomment-919442577

From @tchaikov's [comment](https://github.com/ceph/ceph/pull/41979#discussion_r680537806)
> so each python binding will be built and packaged twice for python3.6 **(or the default python3 on the distro)** and python3.8.

Wouldn't this default version on distro can be bound too, to avoid breaking existing users python binding?

@epuertat @tchaikov 
Atlast, would like to understand if there are plans to drive this PR further, can extend some help if needed.

**Comment by kshtsk** [comment]
The openSUSE recent version is Leap 15.4, which has python3.6 base version, however in stock distro there is python3.9 and even python3.10, but there is no python3.8 at all.

**Comment by cdeshmukh** [comment]
thanks @epuertat 

> For CentOS that's a different story (CentOS 8 doesn't or didn't have all Python packages required for >=3.8, and CentOS 9 is still WIP) (@ktdreyer is that a fair description?)
Are you targetting a specific distro?

We ([cortx-rgw](https://github.com/Seagate/cortx-rgw)) are targeting RockyLinux 8.4 (CentOS 8 equivalent), which has python-3.6 as default python distro. 

So what are the plans for upgrade then for CentOS 8? And this upgrade would be targeted independently for all distros, in ceph build scripts, or need to be together, i believe its the latter case.

**Comment by ktdreyer** [comment]
>So what are the plans for upgrade then for CentOS 8?

[The plan](https://lists.ceph.io/hyperkitty/list/dev@ceph.io/thread/KSHR4QS5X6EMC5V4DTJUJJ4464LX5BPV/#YL7YXVL3IK6L2X6H6GPI6L4JKYIWOJXP) is that `quincy` will be the final release to support CentOS 8, and we'll minimally require CentOS 9 for `reef` and beyond.

**Comment by cdeshmukh** [comment]
> > So what are the plans for upgrade then for CentOS 8?
> 
> [The plan](https://lists.ceph.io/hyperkitty/list/dev@ceph.io/thread/KSHR4QS5X6EMC5V4DTJUJJ4464LX5BPV/#YL7YXVL3IK6L2X6H6GPI6L4JKYIWOJXP) is that `quincy` will be the final release to support CentOS 8, and we'll minimally require CentOS 9 for `reef` and beyond.

thanks for the insightful link for upcoming Reef release, very useful details in there. 

However, i should have been more specific. I meant to ask the plan for 'upgrade to python 3.8/3.9' on CentOS8 itself(until Reef is released & we can adopt), for which this PR was intended originally, along with to support other Distro versions.

**Comment by rkachach** [comment]
@epuertat @tchaikov it's been a while since this PR was closed. Checking the current [supported versions](https://devguide.python.org/versions/) I can see that both `Python 3.6` and `Python 3.7` are already EOL .. 3.8 is also close to its official EOL. Some newer libraries are already dropping support for the EOL versions so probably we have to check the feasibility of bumping up our Python version.

@phlogistonjohn @adk3798 

image


WDYT?

**Comment by tchaikov** [comment]
@rkachach before we drop the support of RHEL 8, we still have to support python 3.6. see also https://github.com/ceph/ceph/pull/56617 .

**Comment by rkachach** [comment]
> Red Hat's Python team will continue to maintain Python 3.6 in RHEL 8 (main EOL is May 31, 2029) and CentOS Stream 8 (EOL is May 31 2024).
> 
> We should consider dropping CentOS 8 support for Ceph in `main`. CentOS 9 ships Python 3.9 as the default Python version.

I see, thanks @tchaikov. @ktdreyer do you know how long should we support RHEL8?

**Comment by adk3798** [comment]
> > Red Hat's Python team will continue to maintain Python 3.6 in RHEL 8 (main EOL is May 31, 2029) and CentOS Stream 8 (EOL is May 31 2024).
> > We should consider dropping CentOS 8 support for Ceph in `main`. CentOS 9 ships Python 3.9 as the default Python version.
> 
> I see, thanks @tchaikov.
> 
> @ktdreyer do you know how long should we support RHEL8?

We are very close to moving over to centos 9 based containers (which have python 3.9 iirc) and that move is considered a blocker for the first squid RC build. On the other hand, we will still be doing quincy and reef releases on centos 8 containers.

**Comment by epuertat** [comment]
> @epuertat do we need to update default value for python3_pkgversion in [ceph.spec.in](https://github.com/ceph/ceph/blob/main/ceph.spec.in?rgh-link-date=2024-07-19T13%3A43%3A11Z#L141) file, since now the minimal supported version of python is 3.9?

In [some RPM docs](https://docs.redhat.com/en/documentation/red_hat_enterprise_linux/9/html/installing_and_using_dynamic_programming_languages/assembly_packaging-python-3-rpms_installing-and-using-dynamic-programming-languages#con_spec-file-description-for-a-python-package_assembly_packaging-python-3-rpms) they suggest that `python3_pkgversion` be the dotted Python version (e.g.: `3.9`, `3.11`), as that's mostly used for core Python packages (`python3.9-devel`) but in our case we're using it for generating the Ceph Python packages (`python3-rados`) and dependencies (`python3-grpcio`).

That might work for installing deps, as most Python packages (that excludes OpenSUSE ecosystem) also define `provides` with the format `pythonX.Y-<package>` (e.g.: `python3.9-grpcio`), but for the package names in the Ceph spec file we should use then another macro. This may mean that the Ceph spec deserves a refactoring to properly deal with Python versions.

**Pull request closed without merging**
